### DISCUSSION TITLE
Add a ParamOptional method

### DISCUSSION THREAD

**Discussion by viccon** [earthboundkid/requests#99] [Ideas] [answered]
Hello,

First and foremost, I want to express my thanks to you for creating this library. I really enjoy using it!

My favourite part of this library is its fluent API. However, there is a scenario that requires me to create intermediary steps, and that is when I'm calling an API that has optional query parameters.

In that case, I have to break the chaining and do empty string/nil checks to not accidentally add empty strings to the query param. I've also experienced some APIs that don't handle those very well.

Therefore, I have a suggestion that I think could improve the DX for these types of situations: the introduction of a `.ParamOptional("key", potentiallyEmptyString)` method.

This method would allow the addition of query parameters which, if their value is an empty string, would be automatically filtered out.

**Comment by earthboundkid** [comment]
Yes, I've wanted this for headers too where it would set the header but only if it's not already set. I'm not totally sure how it should work or how to make it maximum simple and flexible, but it's something I want to do if we can get a good design for it.

**Reply by earthboundkid** [reply]
Maybe "ParamOptional/HeaderOptional sets a query parameter/header to v only if v is non-zero and k is previously unset."?

**Reply by viccon** [reply]
I really like that idea!

**Reply by viccon** [reply]
I could try to implement this if you want!

**Reply by earthboundkid** [reply]
I think I made a branch with this and then got distracted by life stuff. I will try to resurrect it.

**Reply by earthboundkid** [reply]
I started working on this. Still needs more docs and tests, but it works. Got the names backwards tho… https://github.com/earthboundkid/requests/pull/105

**Reply by earthboundkid** [reply, marked as answer]
Fixed the names. Needs an adjustment to the big doc string, but it's close.

**Reply by viccon** [reply]
Awesome!

**Reply by earthboundkid** [reply]
Finally got around to it! Sorry for the long delay.
